src/modules/tools/zprobe/ZGridStrategy.cpp

```diff
@@ -0,0 +1,597 @@
+
+        if (probes == (probe_points-1)){
+            this->cal[X_AXIS] = this->bed_x/2.0f;           // Clear calibration position
+            this->cal[Y_AXIS] = this->bed_y/2.0f;
+            this->cal[Z_AXIS] = this->bed_z/2.0f;                     // Position head for probe removal
+        } else {
+            this->next_cal();                        // to not cause damage to machine due to Z-offset
+        }
+        this->move(this->cal, slow_rate);        // move to the next position
+
+        this->pData[pindex]                      // set offset
+          = z ;                                  // save the offset
+    }
+
+    stream->printf("\nDone!  Please remove probe\n");
+
+    // activate correction
+    this->setAdjustFunction(true);
+
+    this->in_cal = false;
+    return true;
+}
+
+
+void ZGridStrategy::homexyz()
+{
+    Gcode gc("G28", &(StreamOutput::NullStream));
+    THEKERNEL->call_event(ON_GCODE_RECEIVED, &gc);
+}
+
+void ZGridStrategy::move(float *position, float feed)
+{
+    char cmd[64];
+
+    // Assemble Gcode to add onto the queue
+    snprintf(cmd, sizeof(cmd), "G0 X%1.3f Y%1.3f Z%1.3f F%1.1f", position[0], position[1], position[2], feed * 60); // use specified feedrate (mm/sec)
+
+    //THEKERNEL->streams->printf("DEBUG: move: %s\n", cmd);
+
+    Gcode gc(cmd, &(StreamOutput::NullStream));
+    THEKERNEL->robot->on_gcode_received(&gc); // send to robot directly
+}
+
+
+void ZGridStrategy::next_cal(void){
+    if (int(this->cal[X_AXIS] / bed_div_x) % 2 != 0){  // Odd row
+        this->cal[Y_AXIS] -= bed_div_y;
+        if (this->cal[Y_AXIS] < 0){
+            this->cal[X_AXIS] += bed_div_x;
+            if (this->cal[X_AXIS] > bed_x){
+                this->cal[X_AXIS] = 0;
+                this->cal[Y_AXIS] = 0;
+            }
+            else
+                this->cal[Y_AXIS] = 0;
+        }
+    }
+    else {                                          // Even row
+        this->cal[Y_AXIS] += bed_div_y;
+        if (this->cal[Y_AXIS] > bed_y){
+            this->cal[X_AXIS] += bed_div_x;
+            if (this->cal[X_AXIS] > bed_x){
+                this->cal[X_AXIS] = 0;
+                this->cal[Y_AXIS] = 0;
+            }
+            else
+                this->cal[Y_AXIS] = bed_y;
+        }
+    }
+}
+
+
+void ZGridStrategy::setAdjustFunction(bool on)
+{
+    if(on) {
+        // set the compensationTransform in robot
+        THEKERNEL->robot->compensationTransform= [this](float target[3]) { target[2] += this->getZOffset(target[0], target[1]); };
+    }else{
+        // clear it
+        THEKERNEL->robot->compensationTransform= nullptr;
+    }
+}
+
+
+// find the Z offset for the point on the plane at x, y
+float ZGridStrategy::getZOffset(float X, float Y)
+{
+    int xIndex2, yIndex2;
+
+    float xdiff = X / this->bed_div_x;
+    float ydiff = Y / this->bed_div_y;
+
+    float dCartX1, dCartX2;
+
+    int xIndex = (int) xdiff;                  // Get the current sector (X)
```

Comment:
Shouldn't this be rounded up to the next whole number? or is it deliberately truncated to the lowest integer?